lib/src/multires_stochastic_texture_synthesis.rs

```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
+                closest_point_on_boundary_y: y as i32,
+            },
+            ChunkSearchInfo {
+                x: chunk_x,
+                y: chunk_y - 1,
+                center: false,
+                closest_point_on_boundary_x: x as i32,
+                closest_point_on_boundary_y: chunk_y * self.chunk_size as i32 - self.offset_y,
+            },
+            ChunkSearchInfo {
+                x: chunk_x,
+                y: chunk_y + 1,
+                center: false,
+                closest_point_on_boundary_x: x as i32,
+                closest_point_on_boundary_y: (chunk_y + 1) * self.chunk_size as i32 - self.offset_y,
+            },
+            ChunkSearchInfo {
+                x: chunk_x + 1,
+                y: chunk_y + 1,
+                center: false,
+                closest_point_on_boundary_x: (chunk_x + 1) * self.chunk_size as i32 - self.offset_x,
+                closest_point_on_boundary_y: (chunk_y + 1) * self.chunk_size as i32 - self.offset_y,
+            },
+            ChunkSearchInfo {
+                x: chunk_x - 1,
+                y: chunk_y + 1,
+                center: false,
+                closest_point_on_boundary_x: chunk_x * self.chunk_size as i32 - self.offset_x,
+                closest_point_on_boundary_y: (chunk_y + 1) * self.chunk_size as i32 - self.offset_y,
+            },
+            ChunkSearchInfo {
+                x: chunk_x + 1,
+                y: chunk_y - 1,
+                center: false,
+                closest_point_on_boundary_x: (chunk_x + 1) * self.chunk_size as i32 - self.offset_x,
+                closest_point_on_boundary_y: chunk_y * self.chunk_size as i32 - self.offset_y,
+            },
+            ChunkSearchInfo {
+                x: chunk_x - 1,
+                y: chunk_y - 1,
+                center: false,
+                closest_point_on_boundary_x: chunk_x * self.chunk_size as i32 - self.offset_x,
+                closest_point_on_boundary_y: chunk_y * self.chunk_size as i32 - self.offset_y,
+            },
+        ];
+        // Note locking all of them at different times is fine as opposed to
+        // all at once but this behavior should be noted
+        let mut tmp_result: Vec<(i32, i32, i32)> = Vec::with_capacity(k * 9);
+        result.clear();
+        result.reserve(k);
+
+        // this isn't really the kth best distance but should be good enough for how simple it is
+        let mut kth_best_squared_distance = i32::max_value();
+        for place_to_look in places_to_look.iter() {
+            if place_to_look.x >= 0
+                && place_to_look.x < self.grid_width as i32
+                && place_to_look.y >= 0
+                && place_to_look.y < self.grid_height as i32
+            {
+                let is_center = place_to_look.center;
+
+                // a tiny optimization to help us throw far away neighbors
+                // saves us a decent amount of reads
```

Comment:
Is this actually correct though? This seems to be just point-to-point distance against `closest_point_on_boundary_`. The latter seem to be centered on edges of cells, but are not guaranteed to be the closest point inside the cell.

Reply:
Sorry the name there might be a little confusing what I mean by closet point in cell is closest point on boundary. The optimization works as follows:
<img width="166" alt="diagram" src="https://user-images.githubusercontent.com/1163879/71993075-561c6b00-31eb-11ea-8f2a-994480c06614.png">
If the green dot is the pixel we are talking about and we know that the k nearest neighbors are all within the purple circle then we don't have to search any of the red cells because the closest possible points (the boundary points) to our pixel are further than any of our k closest neighbors.

Reply:
Could you not end up in the following situation?

![image](https://user-images.githubusercontent.com/16522064/72115647-17e88e00-3348-11ea-838d-75ba3d1a4a35.png)

I have highlighted in blue what I'm assuming to be the boundary points of the missed cells.

Reply:
Thanks for the picture! I think the position of the blue points is incorrect. They should really be the closest point on each boundary to the current pixel. For the 4 corner cells this point is always a constant (the closest corner on that cell to the center of the grid) no matter which pixel we are addressing in the middle. However for the cells you highlighted as problematic the boundary points I compute are actually based on the green pixels x or y coordinates (in the case of your picture one would be right below the green dot and the other would be directly to the right)

Reply:
Here's a revised version of your picture
<img width="393" alt="diagram_with_projected_points" src="https://user-images.githubusercontent.com/1163879/72120893-43e71d80-330e-11ea-9024-cebcdae82e03.png">

Reply:
Aha, so I did misread the code, thinking you were offsetting from the center. Make sense now, thanks!